## ひろしまQuest2020#stayhome【コース予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 300)

### 管理番号

In [2]:
model_No = 20

### sub-modelを使用するかどうか

In [3]:
use_sub_model = False

In [4]:
ALL_MERGE = 'intermediate/all_merge_{}.f'.format(model_No)
SUBMIT = 'submit/course_{}.csv'.format(model_No)

if use_sub_model:
    ALL_MERGE = 'intermediate/all_merge_{}_sub.f'.format(model_No)
    SUBMIT = 'submit/course_{}_sub.csv'.format(model_No)
else:
    OUT_SUBMODEL = 'intermediate/course_predict_{}.f'.format(model_No)

In [5]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 166)

In [6]:
if use_sub_model:
    all_pitch['predict_curve'] = all_pitch['predict_curve'] / all_pitch['predict_straight']
    all_pitch['predict_slider'] = all_pitch['predict_slider'] / all_pitch['predict_straight']
    all_pitch['predict_shoot'] = all_pitch['predict_shoot'] / all_pitch['predict_straight']
    all_pitch['predict_fork'] = all_pitch['predict_fork'] / all_pitch['predict_straight']
    all_pitch['predict_changeup'] = all_pitch['predict_changeup'] / all_pitch['predict_straight']
    all_pitch['predict_sinker'] = all_pitch['predict_sinker'] / all_pitch['predict_straight']
    all_pitch['predict_cutball'] = all_pitch['predict_cutball'] / all_pitch['predict_straight']
    all_pitch.drop(columns=[
        'predict_0', 'predict_1', 'predict_2', 'predict_3', 'predict_4', 'predict_5', 'predict_6',
        'predict_7', 'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12',
        'predict_straight'
    ], inplace=True)

In [7]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 166)


No  ball  course  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0                  1       1                  1   
1   1   0.0     8.0                  2       1                  1   
2   2   0.0     5.0                  3       1                  1   
3   3   0.0    12.0                  4       1                  2   
4   4   0.0     8.0                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_curve  bc_slider  \
0        0         0           0       0.0     1.0  0.237468   0.325101   
1        0         0           1       0.0     1.0  0.210923   0.247198   
2        0         0           2       0.0     1.0  0.138758   0.130040   
3        1         0           0       0.0     0.0  0.213866   0.474931   
4        1         0           1       0.0     0.0  0.215241   0.420099   

   bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  bc_high_str  \
0  0.209547  0.145227     0.092978   0.027829    0.113620     0.135883   
1  0.177502  0.320359     0.147952   0.039943    0.116772     0.106283   
2  0.092263  0.419906     0.117327   0.038140    0.046858     0.076007   
3  0.139964  0.088519     0.038543   0.007023    0.138168     0.151308   
4  0.145723  0.196762     0.051566   0.009680    0.118444     0.113468   

   bc_high_ball  bc_mid_str  bc_low_str  bc_low_ball  bc_left_str  \
0      0.221489    0.204959    0.174254     0.263415     0.242045   
1      0.230235    0.157397    0.138195     0.367890     0.184621   
2      0.284249    0.102015    0.095055     0.442674     0.118315   
3      0.196937    0.206965    0.157643     0.287147     0.105057   
4      0.206297    0.162901    0.127988     0.389347     0.084346   

   bc_left_ball  bc_center_str  bc_right_str  bc_right_ball  first  second  \
0      0.294634       0.157681      0.115370       0.190270      0       0   
1      0.355720       0.111331      0.105923       0.242405      0       0   
2      0.488095       0.076190      0.078571       0.238828      0       0   
3      0.153718       0.164756      0.246104       0.330367      0       0   
4      0.162167       0.120564      0.199445       0.433477      0       0   

   third  base_cnt  top_bot  role  pos_pit  date_from_opening  elapsed_min  \
0      0         0        0     1        0                  0     0.000000   
1      0         0        0     1        0                  0     0.200000   
2      0         0        0     1        0                  0     0.383333   
3      0         0        0     1        0                  0     1.016667   
4      0         0        0     1        0                  0     1.200000   

   min_diff  total_inning  elapsed_from_inning  ballnum_from_inning  \
0       NaN             0             0.000000                    1   
1  0.200000             0             0.200000                    2   
2  0.183333             0             0.383333                    3   
3  0.633333             0             1.016667                    4   
4  0.183333             0             1.

### train
- 行数: 257117

In [8]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 166)

### test
- 行数: 521650

In [9]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 166)

In [10]:
del all_pitch
gc.collect()

40

In [11]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [12]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

## Dataset作成

In [13]:
lgb_train = lgb.Dataset(train_d, train['course'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [14]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 13,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 8.769293390201968, 
        'lambda_l2': 3.913949617576324e-05, 
        'num_leaves': 6, 
        'feature_fraction': 0.4, 
        'bagging_fraction': 0.8391111798378441, 
        'bagging_freq': 4, 
        'min_child_samples': 50
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 2.33401 + 0.000874282
[200]	cv_agg's multi_logloss: 2.32933 + 0.000982782
[300]	cv_agg's multi_logloss: 2.328 + 0.00108126
[400]	cv_agg's multi_logloss: 2.32743 + 0.00107494
[500]	cv_agg's multi_logloss: 2.32718 + 0.00108418
[600]	cv_agg's multi_logloss: 2.32718 + 0.00112329
Best num_boost_round: 572
Best CV score: 2.3271237073894153
CPU times: user 55min 32s, sys: 11.7 s, total: 55min 44s
Wall time: 6min 59s


In [15]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [16]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 16min 21s, sys: 3.2 s, total: 16min 25s
Wall time: 2min 3s


### Feature Importance

In [17]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
42           elapsed_min       463
136      rank_x_year_bat       465
6    pitcher_cnt_in_game       481
0      pitch_cnt_in_game       482
124           height_bat       499
137      salary_year_bat       547
141         bat_game_cnt       565
117             low_ball       568
50    ave_elapsed_batter       573
114            high_ball       582
47        elapsed_batter       610
119            left_ball       615
122           right_ball       621
125           weight_bat       622
45   elapsed_from_inning       640
115              mid_str       675
139              bmi_bat       684
140           batter_cnt       690
51   diff_elapsed_batter       698
116              low_str       706
43              min_diff       728
53     diff_elapsed_game       737
113             high_str       759
118             left_str       760
48      ave_elapsed_game       788
120           center_str       794
52   diff_elapsed_inning       810
121            right_str       820
49    ave_elapsed_inning       900
41     date_from_opening      1063

In [18]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 4min 12s, sys: 680 ms, total: 4min 13s
Wall time: 32.8 s


## 結果出力

In [19]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.067915  0.148775  0.060944  0.036401  0.061205  0.035715   
1      1  0.049788  0.139140  0.071540  0.030702  0.050656  0.033816   
2      2  0.033319  0.080540  0.059195  0.022743  0.035360  0.024702   
3      3  0.043250  0.112132  0.061633  0.026484  0.036955  0.027757   
4      4  0.049979  0.114146  0.070554  0.037044  0.043122  0.038893   

          6         7         8         9        10        11        12  
0  0.057490  0.054817  0.037231  0.156010  0.085865  0.101389  0.096244  
1  0.047144  0.061052  0.043057  0.150147  0.080982  0.115489  0.126488  
2  0.042169  0.058169  0.033937  0.151553  0.105563  0.200120  0.152630  
3  0.041817  0.057627  0.037042  0.165007  0.099595  0.139021  0.151679  
4  0.049302  0.063072  0.045172  0.135518  0.075114  0.130196  0.147888

In [20]:
submit.to_csv(SUBMIT, header=False, index=False)

### 球種予測で使用する

In [21]:
%%time
if not use_sub_model:
    train_predict = lgb_model.predict(train_d, num_iteration = lgb_model.best_iteration)

CPU times: user 2min 3s, sys: 284 ms, total: 2min 3s
Wall time: 16.2 s


In [22]:
if not use_sub_model:
    df_train_predict = pd.DataFrame(train_predict).reset_index()
    submodel = pd.concat([df_train_predict, submit], ignore_index=True)
    submodel.drop(columns=['index'], inplace=True)
    submodel.rename(columns={
        0: 'predict_0',
        1: 'predict_1',
        2: 'predict_2',
        3: 'predict_3',
        4: 'predict_4',
        5: 'predict_5',
        6: 'predict_6',
        7: 'predict_7',
        8: 'predict_8',
        9: 'predict_9',
        10: 'predict_10',
        11: 'predict_11',
        12: 'predict_12'
    }, inplace=True)
    print(submodel.shape)
    submodel.head()

(778767, 13)


In [23]:
if not use_sub_model:
    submodel.to_feather(OUT_SUBMODEL)

### SIGNATE

In [24]:
'signate submit --competition-id=276 ./{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=276 ./submit/course_20.csv --note feat=166_cv=2.3271237073894153'

### GCS

In [25]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/course_20.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 20(166) ... player同士の組み合わせを削除
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 572
- Best CV score: 2.3271237073894153
- Wall time: 6min 59s
- 評価結果  : 2.3552697

#### 19_sub(197) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 774
- Best CV score: 2.316818574234304
- Wall time: 10min 49s
- 評価結果  : 2.3581029

#### 19(190) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 550
- Best CV score: 2.327580481542456
- Wall time: 7min 36s
- 評価結果  : 2.3554862

#### 18_sub(225) ...サブモデルで球種の予測を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 548
- Best CV score: 2.308116414786115
- Wall time: 8min 25s
- 評価結果  : 2.3638683

#### 18(217) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 373
- Best CV score: 2.3204478805445055
- Wall time: 6min 9s
- 評価結果  : 2.3604751

#### 17(192) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 415
- Best CV score: 2.3200530570975535
- Wall time: 15min 10s
- 評価結果  : 2.3607863

#### 16(181)  ...選手IDごとの2017年のコース実績
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 355
- Best CV score: 2.3203553166835866
- Wall time: 5min 8s
- 評価結果  : 2.3608998

#### 15(153)  ...左右で球種/コースを別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 932
- Best CV score: 2.3387434142884445
- Wall time: 9min 41s
- 評価結果  : 2.3552543 ★

#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648